theano/compile/builders.py

```diff
@@ -168,30 +383,48 @@ def infer_shape(self, node, shapes):
 
         return ret
 
-    def grad(self, inputs, output_grads):
-        if hasattr(self, "grad_ops"):
-            grad_ops = self.grad_ops
-        else:
-            gs = theano.gradient.grad(cost=None,
-                                      known_grads=dict(izip(self.new_outputs,
-                                                            output_grads)),
-                                      wrt=self.new_inputs,
-                                      disconnected_inputs='ignore')
-
-            grad_ops = []
-            for g in gs:
-                if g is None:
-                    grad_ops.append(lambda *args: None)
-                else:
-                    # It is normal if some inputs are not needed in order
-                    # to compute the gradient, so we ignore them.
-                    grad_ops.append(OpFromGraph(self.new_inputs + output_grads,
-                                                [g],
-                                                on_unused_input='ignore'))
-            self.grad_ops = grad_ops
+    def prepare_node(self, node, storage_map, compute_map, impl):
+        if not hasattr(self, "fn") and impl == 'py':
+            self.fn = orig_function(self.local_inputs,
+                                    self.local_outputs,
+                                    **self.kwargs)
+            self.fn.trust_input = True
 
-        return [go(*(inputs + output_grads)) for go in grad_ops]
+    def perform(self, node, inputs, outputs):
+        variables = self.fn(*inputs)
+        assert len(variables) == len(outputs)
+        for output, variable in izip(outputs, variables):
+            # TODO: when function's output-borrowing semantics are correct,
+            # we wont need this copy anymore
+            output[0] = variable.copy()
 
-# Since OpFromGraph contains a Theano compiled function, we should let
-# DebugMode know about it
+
+@gof.local_optimizer([OpFromGraph])
+def inline_ofg_expansion(node):
+    """
+    This optimization expands internal graph of OpFromGraph.
+    Only performed if node.op.is_inline == True
+    Doing so can improve optimization at the cost of compilation speed.
+    """
+    op = node.op
+    if not isinstance(op, OpFromGraph):
+        return False
+    if not op.is_inline:
+        return False
+    return theano.clone(
+        op.local_outputs, {
+            u: v for u, v in izip(
+                node.op.local_inputs, node.inputs)})
+
+optdb.register(
+    'inline_ofg_expansion',
+    gof.opt.in2out(inline_ofg_expansion),
+    0.5, 'fast_compile', 'fast_run')
```

Comment:
Use -0.01 instead of 0.5 and add this comment to explain a little bit the choise:

    We want to run this before the first merge optimizer and before the first scan optimizer.